miner/src/block_assembler.rs

```diff
@@ -413,6 +418,210 @@ impl BlockAssembler {
+            };
+            // try return previous entry
+            if let Some(tx_entry) = entry {
+                if check_tx_skip(&tx_entry) && check_limit_is_ok(&tx_entry) {
+                    return;
+                }
+            }
+            let mut consecutive_failed = 0;
+            for tx_entry in iter {
+                if check_tx_skip(&tx_entry) {
+                    continue;
+                }
+
+                if !check_limit_is_ok(&tx_entry) {
+                    consecutive_failed += 1;
+                    // give up
+                    if consecutive_failed > MAX_CONSECUTIVE_FAILED {
+                        break;
+                    }
+                    continue;
+                }
+
+                // find new tx entry
+                entry.replace(tx_entry.to_owned());
+                return;
+            }
+            // set entry to None if iter is end or consecutive failed
+            entry.take();
+        }
+
+        let mut size: usize = 0;
+        let mut cycles: Cycle = 0;
+        let tx_pool = chain_state.tx_pool();
+        let entries = tx_pool.with_sorted_proposed_txs_iter(|iter| {
+            let mut entries: Vec<TxEntry> = Vec::new();
+            // modified entries, after put a tx into block,
+            // the scores of descendants txs should be updated,
+            // these modified entries is stored in modified_entries.
+            // in each loop,
+            // we pick tx from modified_entries and pool to find the best tx to package
+            let mut modified_entries: TxModifiedEntries = TxModifiedEntries::default();
+            // track in block txs
+            let mut in_block_txs: HashSet<ProposalShortId> = HashSet::new();
+            let mut pool_tx = None;
+            let mut modified_tx = None;
+            loop {
+                // 1. choose best tx from pool and modified entries
+                // 2. compare the two txs, package the better one
+                // 3. update modified entries
+                find_tx_entry(
+                    &mut pool_tx,
+                    iter,
+                    size_limit,
+                    cycles_limit,
+                    size,
+                    cycles,
+                    |tx| {
+                        let tx_id = tx.transaction.proposal_short_id();
+                        in_block_txs.contains(&tx_id) || modified_entries.contains_key(&tx_id)
+                    },
+                );
+                modified_entries.with_sorted_by_score_iter(|iter| {
+                    find_tx_entry(
+                        &mut modified_tx,
+                        iter,
+                        size_limit,
+                        cycles_limit,
+                        size,
+                        cycles,
+                        |_| false,
+                    );
+                });
+                // find tx with higher scores
+                let tx_entry = match (pool_tx.take(), modified_tx.take()) {
```

Comment:
Should implement trait `Ord` for `TxEntry` which delegate to `AncestorsScoreSortKey`, then use `Option#max`, ref: https://doc.rust-lang.org/std/option/enum.Option.html#impl-Ord, no need to use `take` and `replace`

```suggestion
                let tx_entry = [pool_tx, modified_tx].max();
```

Reply:
I'm not sure what happened when None cmp with a Some value.

Reply:
`Option` use `#[derive]` for ordering, doc of the sort order of derived ordering: https://doc.rust-lang.org/std/cmp/trait.PartialOrd.html#derivable, it will return `Ordering::Less` when compare`None` with `Some`